File: erlang/seminars/transport/src/order.erl
-module(order).

-export([start/2]).

-record(ack, {id=na}).
-record(dgr, {id=na, data=[]}).

start(Master, To) ->   {ok, spawn(fun() -> init(Master, To) end)}.

init(Master, To) ->
    receive
	{connect, Net} ->   order(Master, To, 0, 0, [], Net)
    end.

order(Master, To, N, I, [], Net) ->
    receive
	#dgr{id=I, data=Msg} ->
	    Net ! {send, To, #ack{id=I}},	    
	    Master ! Msg,
	    order(Master, To, N, I+1, [], Net);
	#dgr{id=J} when J < I ->
	    Net ! {send, To, #ack{id=J}},	    
	    order(Master, To, N, I, [], Net);
	#ack{} ->
	    order(Master, To, N, I, [], Net);
	{send, Msg} ->
	    Net ! {send, To, #dgr{id=N, data=Msg}},
	    order(Master, To, N+1, I, [{N,Msg}], Net);
	{master, New} ->
	    order(New, To, N, I, [], Net)
    end;
order(Master, To, N, I, [{A,Res}|Rest]=Buffer, Net) ->
    receive
	#dgr{id=I, data=Msg} ->
	    Net ! {send, To, #ack{id=I}},	    
	    Master ! Msg,
	    order(Master, To, N, I+1, Buffer, Net);
	#dgr{id=J} when J < I ->
	    Net ! {send, To, #ack{id=J}},	    
	    order(Master, To, N, I, Buffer, Net);
	#ack{id=A} ->
	    order(Master, To, N, I, Rest, Net);
	#ack{id=B} when B < A ->
	    order(Master, To, N, I, Buffer, Net);
	{send, Msg} ->
	    Net ! {send, To, #dgr{id=N, data=Msg}},
	    order(Master, To, N+1, I, Buffer++[{N,Msg}], Net);
	{master, New} ->
	    order(New, To, N, I, Buffer, Net)
		
    after 10 ->
	    Dgr = #dgr{id=A, data=Res},
	    io:format("order re-sending ~w~n", [Dgr]),
	    Net ! {send, To, Dgr},
	    order(Master, To, N, I, Buffer, Net)
    end.





    
    



	    
			  
	    
    

File: exercises/transport/src/order.ex
defmodule Order do

  require Record

  Record.defrecord(:ord, seq: 0, data: nil)

  Record.defrecord(:ack, seq: 0)
  Record.defrecord(:dgr, seq: 0, data: [])
  
  def start(master, to) do
    {:ok, spawn(fn() -> init(master, to) end)}
  end

  def init(master, to) do
   :io.format("order to ~w: process ~w started~n", [to, self()])
    receive do
      {:connect, netw} ->
	:io.format("order to ~w: connected to ~w~n", [to, netw])
	order(master, to, 0, 0, [], netw)
    end
  end

  def order(master, to, n, i, [], netw) do
    receive do
	
      ord(seq: ^i, data: msg) ->
	send(netw, {:send, to, ack(seq: i)})
	send(master, msg)
	order(master, to, n, i+1, [], netw)
	
     ord(seq: j) when j < i ->
	send(netw, {:send, to, ack(seq: j)})
	order(master, to, n, i, [], netw)

      ack() ->
	order(master, to, n, i, [], netw)

      {:send, msg} ->
	send(netw, {:send, to, ord(seq: n, data: msg)})
	order(master, to, n+1, i, [{n,msg}], netw);

      {:master, new} ->
	order(new, to, n, i, [], netw)
    end
  end
  def order(master, to, n, i, [{a,res}|rest]=buffer, netw) do
    receive do

      ord(seq: ^i, data: msg) ->
	send(netw, {:send, to, ack(seq: i)})
	send(master, msg)
	order(master, to, n, i+1, buffer, netw)

      ord(seq: j) when j < i ->
	send(netw, {:send, to, ack(seq: j)})
	order(master, to, n, i, buffer, netw)

      ack(seq: ^a) ->
	order(master, to, n, i, rest, netw)

      ack(seq: b) when b < a ->
	order(master, to, n, i, buffer, netw);

      {:send, msg} ->
	send(netw,{:send, to, ord(seq: n, data: msg)})
	order(master, to, n+1, i, buffer++[{n,msg}], netw)

      {:master, new} ->
	order(new, to, n, i, buffer, netw)
		
    after 10 ->
	dgr = ord(seq: a, data: res)
	##:io.format("order to ~w resending ~w\", [to, dgr])
	send(netw, {:send, to, dgr})
	order(master, to, n, i, buffer, netw)
    end
  end
end





    
    



	    
			  
	    
    
